src/main/java/org/opensearch/timeseries/transport/ResultProcessor.java

```diff
@@ -309,19 +320,15 @@ public void onResponse(CompositeRetriever.Page entityFeatures) {
                     } catch (Exception e) {
                         LOG.error("Unexpected exception", e);
                         handleException(e);
+                    } finally {
+                        // Decrement pagesInFlight after processing is complete
+                        pagesInFlight.decrementAndGet();
                     }
                 });
-            }
-
-            if (!pageIterator.hasNext() && config.getImputationOption() != null) {
-                if (sentOutPages.get() > 0) {
-                    // at least 1 page sent out. Wait until all responses are back.
-                    scheduleImputeHCTask();
-                } else {
-                    // no data in current interval. Send out impute request right away.
-                    imputeHC(dataStartTime, dataEndTime, configId, taskId);
-                }
-
+            } else {
+                // No entity features to process
+                // Decrement pagesInFlight immediately
+                pagesInFlight.decrementAndGet();
```

Comment:
Just to confirm, here every page is from the feature aggregation results, we want to make sure we received every page and then sending it out to processing (means sending the aggregated feature data to the correct model and doing .process())? then after we check each entity if data was received and send impute call to place the imputed value?

Reply:
yes